pkg/optimizer/flatzinc_model.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
+
+package optimizer
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"strings"
+)
+
+type FlatZincParam struct {
+	NumInstances uint // NumInstances > 1 means an array of params
+	Name         string
+	Type         string
+	Assignment   string
+}
+
+func (fzp FlatZincParam) ParamDeclaration() string {
+	if fzp.NumInstances == 1 {
+		return fmt.Sprintf("%s: %s = %s;\n", fzp.Type, fzp.Name, fzp.Assignment)
+	}
+	return fmt.Sprintf("array [1..%d] of %s: %s = %s;\n", fzp.NumInstances, fzp.Type, fzp.Name, fzp.Assignment)
+}
+
+type Annotations []string
+
+func (annots Annotations) AnnotationString() string {
+	annotsStr := strings.Join(annots, " :: ")
+	if len(annotsStr) > 0 {
+		annotsStr = " :: " + annotsStr
+	}
+	return annotsStr
+}
+
+type FlatZincVariable struct {
+	NumInstances uint // NumInstances > 1 means an array of vars
+	Name         string
+	Type         string
+	Assignment   string
+	Annotations
+}
+
+func (fzv FlatZincVariable) VarDeclaration() string {
+	if fzv.NumInstances == 1 {
+		return fmt.Sprintf("var %s: %s%s;\n", fzv.Type, fzv.Name, fzv.AnnotationString())
+	}
+	return fmt.Sprintf("array [1..%d] of var %s: %s%s;\n", fzv.NumInstances, fzv.Type, fzv.Name, fzv.AnnotationString())
+}
+
+type FlatZincConstraint struct {
+	Identifier  string
+	Expressions []string
+	Annotations
+}
+
+func (cnstr FlatZincConstraint) ConstraintStatement() string {
+	exprs := strings.Join(cnstr.Expressions, ", ")
+	return fmt.Sprintf("constraint %s(%s)%s;\n", cnstr.Identifier, exprs, cnstr.AnnotationString())
+}
+
+type SolveGoal int64
+
+const (
+	Satisfy SolveGoal = iota
+	Minimize
+	Maximize
+)
+
+func (s SolveGoal) String() string {
+	switch s {
+	case Satisfy:
+		return "satisfy"
+	case Minimize:
+		return "minimize"
+	case Maximize:
+		return "maximize"
+	}
+	return "unknown"
+}
```

Comment:
I believe we'll need the reverse function as well - given the directives received from the policies, determine the SolveGoal value.

Reply:
Will the directives received from the policies use the same keywords?

Reply:
This is the suggested definition (can be changed if needed):
```
type OptimizationDirective string

// List of directives
const (
	Minimize OptimizationDirective = "min"
	Maximize OptimizationDirective = "max"
)
```